### PR TITLE
Feature returns required purposes

### PR DESCRIPTION
**Pull request by jonathangoulding**
Due to the estimated complexity of adding fetch services for purposes, points and other display items. We have broken down the complex display items into their own branches of work.

https://eaflood.atlassian.net/browse/WATER-4386

This branch will add purposes to the return requirement summary card

Previous work done here - https://github.com/DEFRA/water-abstraction-system/pull/1019

### CODE REVIEW

File: app/services/return-requirements/check.service.js [hunk 1]
```diff
@@ -8,6 +8,8 @@
 const CheckPresenter = require('../../presenters/return-requirements/check.presenter.js')
 const SessionModel = require('../../models/session.model.js')
 
+const RequirementsService = require('./check/returns-requirements.service.js')
```
Comment (jonathangoulding) [inline app/services/return-requirements/check.service.js:11]: Rename returns requirements

File: app/services/return-requirements/check.service.js [hunk 2]
```diff
@@ -21,13 +23,16 @@ async function go (sessionId, yar) {
 
   await _markCheckPageVisited(session)
 
+  const requirements = await RequirementsService.go(sessionId)
```
Comment (jonathangoulding) [inline app/services/return-requirements/check.service.js:26]: Rename returns requirements ?

File: app/services/return-requirements/check/returns-requirements.presenter.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+const { formatAbstractionDate } = require('../../../presenters/base.presenter.js')
+
+/**
+ * Formats data for the `/return-requirements/{sessionId}/check` page
+ * @module ReturnRequirementsPresenter
+ */
+
+function go (requirements, purposeIds, journey) {
+  //  Not clear of this can be an if else
```
Comment (jonathangoulding) [inline app/services/return-requirements/check/returns-requirements.presenter.js:11]: Check is this can be assumed ?

File: app/services/return-requirements/check/returns-requirements.service.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
+'use strict'
+
+/**
+ * Orchestrates fetching and presenting the requirements for the check page
```
Comment (jonathangoulding) [inline app/services/return-requirements/check/returns-requirements.service.js:4]: returns requirements

File: app/services/return-requirements/check/returns-requirements.service.js [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * @module RequirementsService
```
Comment (jonathangoulding) [inline app/services/return-requirements/check/returns-requirements.service.js:5]: ReturnRrequirements

File: app/services/return-requirements/check/returns-requirements.service.js [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+const FetchPurposesByIdsService = require('./fetch-purposes.service.js')
+const ReturnRequirementsPresenter = require('./returns-requirements.presenter.js')
+const SessionModel = require('../../../models/session.model.js')
+
+/**
+ * Orchestrates fetching and presenting the requirements for `/return-requirements/{sessionId}/check` page
```
Comment (jonathangoulding) [inline app/services/return-requirements/check/returns-requirements.service.js:13]: return

File: app/services/return-requirements/check/returns-requirements.service.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * @param {string} sessionId - The UUID for return requirement setup session record
+ *
+ * @returns {Promise<Object>} page data needed by the view template
+ */
+async function go (sessionId) {
+  const session = await SessionModel.query().findById(sessionId)
```
Comment (jonathangoulding) [inline app/services/return-requirements/check/returns-requirements.service.js:20]: Should this be lifted to it's own fetch ?

Comment (jonathangoulding) [reply app/services/return-requirements/check/returns-requirements.service.js:20]: From previous code

File: app/services/return-requirements/check/returns-requirements.service.js [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+function _purposeIds (requirements) {
+  const requirementPurposes = requirements.flatMap((requirement) => {
+    if (requirement.purposes) {
+      return requirement.purposes
+    }
+
+    return []
+  })
+
+  //  Duplicate is a bug this is temp
```
Comment (jonathangoulding) [inline app/services/return-requirements/check/returns-requirements.service.js:38]: We are seeing duplicate purposes in the view and subsequently getting duplicates in the array when selected.

File: app/views/return-requirements/check.njk [hunk 9]
```diff
@@ -27,81 +34,81 @@
...
           classes: 'govuk-summary-list govuk-summary-list__row--no-border',
           key: {
-            text: "Start date"
-          },
+          text: "Start date"
+        },
           value: {
-            html: '<span data-test="meta-data-start-date">' + startDate + '</span>'
-          },
+          text: startDate
```
Comment (jonathangoulding) [inline app/views/return-requirements/check.njk:40]: Where has this come from ?

File: app/views/return-requirements/check.njk [hunk 10]
```diff
@@ -27,81 +34,81 @@
...
         {
           classes: 'govuk-summary-list govuk-summary-list__row--no-border',
           key: {
-            text: "Reason"
-          },
+          text: "Reason"
+        },
           value: {
-            html: '<span data-test="meta-data-reason">' + reason + '</span>'
-          },
```
Comment (jonathangoulding) [inline app/views/return-requirements/check.njk:52]: Have these come from a previous pr ?

File: app/views/return-requirements/check.njk [hunk 11]
```diff
@@ -110,152 +117,152 @@
       </form>
       {% for requirement in requirements %}
```
Comment (jonathangoulding) [inline app/views/return-requirements/check.njk:118]: Does this assume the returns requirement has a length ? Guard needed ?

File: app/views/return-requirements/check.njk [hunk 12]
```diff
@@ -314,7 +321,7 @@
       <hr class="govuk-section-break govuk-section-break--xl govuk-section-break--visible">
 
       {{ govukButton({
-        html: '<span data-test="meta-data-approve">' + "Approve returns requirements" + "</span>",
```
Comment (jonathangoulding) [inline app/views/return-requirements/check.njk:317]: same as others

File: app/services/return-requirements/check/returns-requirements.service.js [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Orchestrates fetching and presenting the return requirements for `/return-requirements/{sessionId}/check` page
+ *
+ * @param {string} sessionId - The UUID for return requirement setup session record
+ *
+ * @returns {Promise<Object>} page data needed by the view template
+ */
+async function go (sessionId) {
+  const session = await SessionModel.query().findById(sessionId)
```
Comment (Cruikshanks) [inline app/services/return-requirements/check/returns-requirements.service.js:20]: As this is called by `CheckService` which has already fetched the `SessionModel` instance we can pass that in instead of the ID and avoid making the query again.

I appreciate every other service in return requirements fetch the session. But this is not because they are following a convention for services. Simply the other pages are simple enough a single service is all that is needed.

File: app/services/return-requirements/check/returns-requirements.service.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
+'use strict'
+
+/**
+ * Orchestrates fetching and presenting the return requirements for the check page
+ * @module ReturnRequirementsService
+ */
+
+const ReturnRequirementsPresenter = require('../../../presenters/return-requirements/check/returns-requirements.presenter.js')
+const SessionModel = require('../../../models/session.model.js')
+const PurposeModel = require('../../../models/purpose.model.js')
```
Comment (Cruikshanks) [inline app/services/return-requirements/check/returns-requirements.service.js:10]: NIT - Alphabetical ordering

```suggestion
const PurposeModel = require('../../../models/purpose.model.js')
const ReturnRequirementsPresenter = require('../../../presenters/return-requirements/check/returns-requirements.presenter.js')
const SessionModel = require('../../../models/session.model.js')
```

File: app/services/return-requirements/check/returns-requirements.service.js [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+async function _fetchPurposeIds (purposeIds) {
+  return PurposeModel.query()
+    .select([
+      'purposes.id',
+      'purposes.description'
+    ])
+    .findByIds(purposeIds)
+}
```
Comment (Cruikshanks) [inline app/services/return-requirements/check/returns-requirements.service.js:48]: NIT - alphabetical ordering

```suggestion
async function _fetchPurposeIds (purposeIds) {
  return PurposeModel.query()
    .select([
      'purposes.id',
      'purposes.description'
    ])
    .findByIds(purposeIds)
}

function _purposeIds (requirements) {
  const requirementPurposes = requirements.flatMap((requirement) => {
    if (requirement.purposes) {
      return requirement.purposes
    }

    return []
  })

  return [...new Set(requirementPurposes)]
}
```

File: app/views/return-requirements/check.njk [hunk 16]
```diff
@@ -27,81 +34,81 @@
...
-            ]
-          }
+          items: [
+            {
+              href: reasonLink,
+              text: "Change",
+              visuallyHiddenText: "the reason for the return requirement"
+            }
+          ]
+        }
```
Comment (Cruikshanks) [inline app/views/return-requirements/check.njk:68]: A bunch of indentations seem to have been removed that I think were valid. Was this intentional or some auto-formatter?

File: app/views/return-requirements/check.njk [hunk 17]
```diff
@@ -27,81 +34,81 @@
...
+            {
+              href: "note",
+              text: "Change" if note else "Add a note"
+            },
+            {
+              href: "delete-note",
+              text: "Delete"
+            } if note
+          ]
+        }
```
Comment (Cruikshanks) [inline app/views/return-requirements/check.njk:101]: Same thing about indentations.

File: app/views/return-requirements/check.njk
Comment (Cruikshanks) [inline app/views/return-requirements/check.njk:1]: Added a couple of comments before seeing that changes seem to have been made throughout to the indentation.

It doesn't look right. I'm not sure if it was intentional or automatic but I'm happy to discuss if it was intended.

**Review by Cruikshanks — CHANGES_REQUESTED** [review]
Just a couple of NITs, highlighting something has happened with the indentation in the view, and a suggestion about restructuring the tests slightly.

File: test/services/return-requirements/check/returns-requirements.service.test.js [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+        frequencyReported: 'daily',
+        index: 0,
+        purposes: [
+          'Spray Irrigation - Storage'
+        ],
+        siteDescription: 'A place in the sun'
+      }],
+      returnsRequired: true
+    })
+  })
```
[the remark this answers is not in the document]
Comment (jonathangoulding) [reply test/services/return-requirements/check/returns-requirements.service.test.js]: Addressed. 

Session stripped back the required data and the call removed to the model

File: app/services/return-requirements/check/returns-requirements.service.js (hunk 13 above)
Comment (jonathangoulding) [reply app/services/return-requirements/check/returns-requirements.service.js]: Have addressed and updated the tests

File: app/views/return-requirements/check.njk [hunk 19]
```diff
@@ -110,202 +117,204 @@
...
+          }) }}
         </div>
       {% endfor %}
     </div>
   {% endif %}
 
   <form method="post">
     <div class="govuk-body">
-      {% if journey == 'no-returns-required' %}
+      {% if noReturnsRequired %}
```
[the remark this answers is not in the document]
Comment (jonathangoulding) [reply app/views/return-requirements/check.njk:283]: Decisions yours. 

The journey is always one instance of a string as far as i can see. 

I assumed it would be safe to just the the returnsRequired flag. But u figured this must be done for a reason.

File: app/views/return-requirements/check.njk
Comment (jonathangoulding) [reply app/views/return-requirements/check.njk:1]: Will fix indentation.

File: app/views/return-requirements/check.njk (hunk 19 above)
[the remark this answers is not in the document]
Comment (Cruikshanks) [reply app/views/return-requirements/check.njk:283]: It is always one instance. So, I think its fine/handy to convert that to the boolean value `returnsRequired` in the presenter and use that in the view. But when it comes to the view we can just do `{% if !returnsRequired %}`.

File: app/views/return-requirements/check.njk [hunk 20]
```diff
@@ -309,12 +314,16 @@
...
+        {% else %}
+          <h3 class="govuk-heading-m">Returns are not required for this licence</h3>
+          <hr class="govuk-!-margin-bottom-5 govuk-section-break--s govuk-section-break--visible">
       {% endif %}
 
       <hr class="govuk-section-break govuk-section-break--xl govuk-section-break--visible">
 
       {{ govukButton({
         html: '<span data-test="meta-data-approve">' + "Approve returns requirements" + "</span>",
+        text: "Approve returns requirements",
```
Comment (Cruikshanks) [inline app/views/return-requirements/check.njk:326]: I could have sworn I flagged this on the previous PR!

Anyway, according to [the docs](https://design-system.service.gov.uk/components/button/) if the `html:` option is provided `text:` will be ignored. However, in this case we don't need to provide a `html:` option. Someone has done this to give it a Cypress test attribute, but we don't need these for buttons.

I'd delete the `html:` option and keep the `text:` you added.

File: test/presenters/return-requirements/check/returns-requirements.presenter.test.js [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+          ],
+          siteDescription: 'A place in the sun'
+        }]
+      })
+    })
+  })
+
+  describe("the 'requirements' property", () => {
+    describe('when the requirement has agreements exceptions', () => {
+      it('correctly returns and requirement with agreements exceptions', () => {
```
Comment (Cruikshanks) [inline test/presenters/return-requirements/check/returns-requirements.presenter.test.js:74]: Ahhh, I understand why this is described as such. The presenter is checking for agreements exceptions being present as an indicator the requirement is 'complete'. So, i'd suggest rewording based on this.

```suggestion
    describe('when the requirement is "complete" (agreements exceptions is populated)', () => {
      it('correctly returns the requirement', () => {
```

File: test/presenters/return-requirements/check/returns-requirements.presenter.test.js [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+          index: 0,
+          purposes: [
+            'A singular purpose'
+          ],
+          siteDescription: 'A place in the sun'
+        }
+        ])
+      })
+    })
+    describe('when the requirement does not have any agreements exceptions', () => {
```
Comment (Cruikshanks) [inline test/presenters/return-requirements/check/returns-requirements.presenter.test.js:90]: NIT - whitespace

Plus an update to the wording

```suggestion
    })

    describe('when the requirement is "incomplete" (agreements exceptions is not populated)', () => {
```

File: test/presenters/return-requirements/check/returns-requirements.presenter.test.js [hunk 23]
```diff
@@ -0,0 +1,123 @@
...
+        delete requirements[0].agreementsExceptions
+      })
+
+      it('does not return the requirement', () => {
+        const result = ReturnRequirementsPresenter.go(requirements, purposeIds, journey)
+
+        expect(result.requirements).to.equal([])
+      })
+    })
+    describe('has purpose id\'s', () => {
```
Comment (Cruikshanks) [inline test/presenters/return-requirements/check/returns-requirements.presenter.test.js:101]: NIT - whitespace

But, I'd also suggest moving this test into the `describe('when the requirement is "complete" (agreements exceptions is populated)')` block as we can assume a completed requirement will have purposes.

So, you'd end up with

```javascript
describe('when the requirement is "complete" (agreements exceptions is populated)', () => {
  it('correctly returns the requirement', () => {
  })

  it('maps the selected purpose ID\'s to their descriptions', () => {
  })
})
```

**Review by Cruikshanks — CHANGES_REQUESTED** [review]

File: app/views/return-requirements/check.njk (hunk 20 above)
Comment (jonathangoulding) [reply app/views/return-requirements/check.njk:326]: I was wondering why this was an issue. I must have messed up the initial branch.

File: app/presenters/return-requirements/check/returns-requirements.presenter.js [hunk 24]
```diff
@@ -0,0 +1,60 @@
+'use strict'
+
+const { formatAbstractionDate } = require('../../base.presenter.js')
+
+/**
+ * Formats return requirements data for the `/return-requirements/{sessionId}/check` page
+ * @module ReturnRequirementsPresenter
+ */
+
+function go (requirements, purposeIds, journey) {
```
[the remark this answers is not in the document]
Comment (jonathangoulding) [reply app/presenters/return-requirements/check/returns-requirements.presenter.js:10]: Much better will make the change

[the remark this answers is not in the document]
Comment (jonathangoulding) [reply app/presenters/return-requirements/check/returns-requirements.presenter.js]: Addressed.

File: app/presenters/return-requirements/check/returns-requirements.presenter.js [hunk 25]
```diff
@@ -5,14 +5,17 @@ const { formatAbstractionDate } = require('../../base.presenter.js')
 /**
  * Formats return requirements data for the `/return-requirements/{sessionId}/check` page
  * @module ReturnRequirementsPresenter
+ * @param {object} requirements - The requirements from the session
+ * @param {module:PurposeModel} purposes - The purposes related to the ids of every requirement purpose
+ * @param {string} journey - The journey from the session to determine if the journey is returns required
+ *
+ * @returns {Object} returns requirement data needed by the view template
```
Comment (Cruikshanks) [inline app/presenters/return-requirements/check/returns-requirements.presenter.js:12]:
```suggestion
```

File: app/services/return-requirements/check/returns-requirements.service.js [hunk 26]
```diff
@@ -11,26 +11,26 @@ const ReturnRequirementsPresenter = require('../../../presenters/return-requirem
...
 
   const purposeIds = _purposeIds(requirements)
-  const purposes = await _fetchPurposeIds(purposeIds)
+  const purposes = await _fetchPurposes(purposeIds)
 
   return ReturnRequirementsPresenter.go(requirements, purposes, journey)
 }
 
-async function _fetchPurposeIds (purposeIds) {
+async function _fetchPurposes (purposes) {
```
Comment (Cruikshanks) [inline app/services/return-requirements/check/returns-requirements.service.js:27]:
```suggestion
async function _fetchPurposes (purposeIds) {
```

File: app/services/return-requirements/check/returns-requirements.service.js [hunk 27]
```diff
@@ -11,26 +11,26 @@ const ReturnRequirementsPresenter = require('../../../presenters/return-requirem
...
   return PurposeModel.query()
     .select([
       'purposes.id',
       'purposes.description'
     ])
-    .findByIds(purposeIds)
+    .findByIds(purposes)
```
Comment (Cruikshanks) [inline app/services/return-requirements/check/returns-requirements.service.js:33]:
```suggestion
    .findByIds(purposeIds)
```

File: app/views/return-requirements/check.njk
Comment (Cruikshanks) [inline app/views/return-requirements/check.njk:1]: Eeek! We need these in the summary list in order for Cypress to find these elements and test them.

It was the `html:` property in the _button_ that can be deleted.

File: test/presenters/return-requirements/check/returns-requirements.presenter.test.js (hunk 23 above)
Comment (Cruikshanks) [reply test/presenters/return-requirements/check/returns-requirements.presenter.test.js]: Unless someone maliciously mucks about with the order that the journey is completed (in which case we blow up and on their head be it!) we will only call the functionality in the presenter to map purposes for a 'complete' requirement.

This means that code will _never_ be asked to format an empty array of purposes. So, I don't think there is any value in testing it. I've also double-checked the test output, and it shows no missing coverage for this presenter, so code coverage won't be a problem either.

File: app/services/return-requirements/check/returns-requirements.service.js (hunk 26 above)
Comment (jonathangoulding) [reply app/services/return-requirements/check/returns-requirements.service.js:27]: I did spot and fix this on the other branch just now !

File: app/views/return-requirements/check.njk
Comment (jonathangoulding) [reply app/views/return-requirements/check.njk:1]: Addressed will let you resolve.

**Review by Cruikshanks — APPROVED** [review]
I think that's it!
